### PR TITLE
Story branch

### PR DESCRIPTION
**Pull request by jaholt**
The edit, delete and add to database functionality. Commit message has more detail.

### CODE REVIEW

File: index.php [hunk 1]
```diff
@@ -1,5 +1,6 @@
 <?php
-$db = new PDO('MySQL: host = 127.0.0.1;dbname=portfolioProject', 'root');
+
```
Comment (charliecog) [inline index.php:2]: copy everything from index.html into here

Comment (charliecog) [reply index.php:2]: and get rid of index.html

File: functions.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+/*
+ * Is used to select data from the database and return it as an array
+ *
+ * @param1 - PDO - this is where the $db var should be used otherwise the function can't access the database(scope)
+ * @param2 - $table - used to choose which table you want from the db, should be used in double quotes with backticks.
+ * @param3 - $field - which field you want from the table, in double quotes with backticks
+ *
+ * @return - an array - should be an array of the field that's picked.
+ */
+function retrieveData ($db, $table, $field) :array
```
Comment (charliecog) [inline functions.php]: type hint params

File: adminPage/adminPage.php [hunk 3]
```diff
@@ -1,22 +1,22 @@
 <?php
-require '../dbConn.php';
-require 'adminPageActions.php';
+require_once '../dbConn.php';
+require_once 'adminPageActions.php';
+require_once '../functions.php';
 
+$projectsArray = retrieveData($db, "`projects`", "*");
```
Comment (charliecog) [inline adminPage/adminPage.php]: shouldnt use select *

File: adminPage/adminPage.php [hunk 4]
```diff
@@ -1,22 +1,22 @@
...
 foreach ($projectsArray as $project) {
```
Comment (charliecog) [inline adminPage/adminPage.php]: make a function out of this

File: adminPage/adminPageActions.php [hunk 5]
```diff
@@ -1,38 +1,25 @@
...
-}
-$projectsArray = retrieveProjectsData($db);
-
 $projectsDeleteId = $_POST['deleteProjectId'];
-$projectsDeleteQuery = $db->prepare("DELETE FROM `projects` WHERE `projects`.`id` = :deleteId;");
-$projectsDeleteQuery->bindParam(':deleteId',$projectsDeleteId);
-$projectsDeleteQuery->execute();
 
-////comments here
+$aboutMeText = $_POST['aboutMeText'];
```
Comment (charliecog) [inline adminPage/adminPageActions.php]: should check if individual parts of POST data exist before assigning them to variables.

File: adminPage/adminPageActions.php [hunk 6]
```diff
@@ -1,38 +1,25 @@
...
+$techText = $_POST['techText'];
+$prevExpText = $_POST['prevExpText'];
+$otherExpText = $_POST['otherExpText'];
+
+insertProject($db, $projectsTitle, $projectsText, $projectsUrl);
```
Comment (charliecog) [inline adminPage/adminPageActions.php]: try to create logic that only call these functions if they are needed not just all of them by default

File: adminPage/adminPageActions.php [hunk 7]
```diff
@@ -1,38 +1,25 @@
...
+
+deleteProject($db, $projectsDeleteId);
+
+replaceAboutMe($db, $aboutMeText, $techText, $prevExpText, $otherExpText);
+
+
 //function retrieveAboutMeData($db) {
```
Comment (charliecog) [inline adminPage/adminPageActions.php]: remove commented out code

File: functions.php [hunk 8]
```diff
@@ -0,0 +1,98 @@
+<?php
+require_once 'dbConn.php';
+
```
Comment (charliecog) [inline functions.php:3]: remove line break

File: functions.php [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    $result = $query->fetchAll();
+    if (is_array($result) == true) {
+        return $result;
+    }
+}
+
+///*
+// * Is used to turn the array result into text that can be displayed on a page
+// *
+// * @param1 - array - this should be the array that you want to be text
```
Comment (charliecog) [inline functions.php]: delete commented out code

File: functions.php [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+//        return "$array[$key]";
+//    } else {
+//        return 'failed to turn into text';
+//    }
+//}
+
+/*
+ * This function is used to add a new project to the db
+ *
+ * @param1 - PDO - this is where the database variable will be inputted into the array
```
Comment (charliecog) [inline functions.php]: take a look at how to doc block, need to identify which param by name

File: functions.php [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+ * @return - int - the no. of rows, truthy to know if the function worked
+ */
+function insertProject ($db, $title, $text, $url)
+{
+    $projectsInsertQuery = $db->prepare("INSERT INTO `projects` (`title`, `fillText`, `url`) VALUES (:projectsTitle, :projectsText, :projectsUrl)");
+    $projectsInsertQuery->bindParam(':projectsTitle', $title);
+    $projectsInsertQuery->bindParam(':projectsText', $text);
+    $projectsInsertQuery->bindParam(':projectsUrl', $url);
+    $projectsInsertQuery->execute();
+    return $projectsInsertQuery;
```
Comment (charliecog) [inline functions.php:58]: can return $projectsInsertQuery->execute(); 
and save a line

File: functions.php [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+/*
+ * used to delete a porject from inside the database
```
Comment (charliecog) [inline functions.php]: project

File: functions.php [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+ *
+ * @param1 - PDO - the $db variable should be used here
+ * @parma2 - int - a number that corresponds to the id of the row which the project is stored on, can be worked out by the data displayed at the top of the page
+ *
+ * @return - int - the no. of rows effected, truthy to know if the function worked
+ */
+function deleteProject($db, $projectId)
+{
+    $projectsDeleteQuery = $db->prepare("DELETE FROM `projects` WHERE `projects`.`id` = :deleteId;");
```
Comment (charliecog) [inline functions.php]: `id` is enough, dont need `projects`.`id`

**Review by charliecog — COMMENTED** [review]
looking good functionality wise, just a few bits to pick up with how the code is laid out.

File: index.php (hunk 1 above)
Comment (jaholt) [reply index.php:2]: Done, although still have html so I don't start breaking my site navigation just yet

File: functions.php (hunk 2 above)
Comment (jaholt) [reply functions.php]: done

File: functions.php (hunk 8 above)
Comment (jaholt) [reply functions.php:3]: done

File: functions.php (hunk 9 above)
Comment (jaholt) [reply functions.php]: I need it later when it works properly

File: adminPage/adminPageActions.php (hunk 7 above)
Comment (jaholt) [reply adminPage/adminPageActions.php]: no

File: adminPage/adminPage.php (hunk 3 above)
Comment (jaholt) [reply adminPage/adminPage.php]: but i want to select everything

File: functions.php (hunk 10 above)
Comment (jaholt) [reply functions.php]: done

File: functions.php (hunk 12 above)
Comment (jaholt) [reply functions.php]: done

File: functions.php (hunk 13 above)
Comment (jaholt) [reply functions.php]: done

**Comment by olivergate** [comment]
admin.html has to go.

File: aboutMePage/aboutMePage.php [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    <link rel="stylesheet" href="styles.css" type="text/css" />
+</head>
+<body>
+<nav>
+    <div class="navContainer">
+        <div class="logoName">
+            JaHolt
+        </div>
+        <ul>
+            <a href="../contactPage/contactPage.php"><li>CONTACT</li></a>
```
Comment (olivergate) [inline aboutMePage/aboutMePage.php:20]: You shouldn't be putting a tags around entire elements. put the a tag inside the <li> element

File: adminPage/adminPageActions.php [hunk 15]
```diff
@@ -1,58 +1,27 @@
...
-function retrieveProjectsData($db) {
-    $db->setAttribute(PDO::ATTR_DEFAULT_FETCH_MODE, PDO::FETCH_ASSOC);
-    $projectsEditQuery = $db->prepare("SELECT * FROM `projects`;");
-    $projectsEditQuery->execute();
-    $projectsResult = $projectsEditQuery->fetchAll();
-    return $projectsResult;
+require_once '../dbConn.php';
+require_once '../functions.php';
+
+if($_POST['portfolioTitle'] && $_POST['projectText'] && $_POST['projectUrl']){
```
Comment (olivergate) [inline adminPage/adminPageActions.php]: PSR2 violation. 
If statements with multiple conditoin should be set on different lines like
if ( 
      condition1
     conditoin2
) {

File: functions.php [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+ * This function is designed to turn the arrays from the projects table into text and html to be displayed on the portfolio page
+ *
+ * @param array $projectResult this param will be the multidimensional array retrieved from the projects table assigned to a variable that equals the retrieveProjectData function
+ *
+ * @return string returns the array parcelled up into html that can be displayed in the page and acted on by the pre-existing css
+ */
+function produceProject (array $projectResult) : string
+{
+    $result = '';
+    if (is_array($projectResult) && array_key_exists(0, $projectResult)) {
```
Comment (olivergate) [inline functions.php]: Same issue as above

File: index.php (hunk 1 above)
Comment (olivergate) [reply index.php:2]: Get rid of unnecessary files.

File: tests/functionTests.php [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+     * This success test checks whether the function returns a string of html
+     */
+    public function testPrintAdminProjectsFailure()
+    {
+        $expected = 'the projects array has not been entered correctly';
+        $example = ['id' => 'id', 'title' => 'title', 'fillText' => 'fillText', 'url' => 'url'];
+        $case = printAdminProjects($example);
+        $this->assertEquals($case, $expected);
+    }
+}
```
Comment (olivergate) [inline tests/functionTests.php:77]: Get rid of uneccesarry SQL files

**Pull request merged**